---
title: "Your Caculated PGS"
root.dir: "/Dedicated/jmichaelson-wdata/msmuhammad/scratch/PGS/src"
author: "Muhammad Elsadany"
date: "`r Sys.Date()`"
output:
  html_document:
    code_folding: hide
  pdf_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE, fig.align='center')
rm(list = ls())
gc()
source("/Dedicated/jmichaelson-wdata/msmuhammad/msmuhammad-source.R")
formatted_tbl <- function(x) {
  kableExtra::kable_styling(knitr::kable(x))
}
library(plotly)
project.dir <- "/Dedicated/jmichaelson-wdata/msmuhammad/scratch/PGS"
setwd(project.dir)
knitr::opts_knit$set(root.dir = project.dir)
```



```{r pgs combining}
pgs.dir <- "/Dedicated/jmichaelson-wdata/msmuhammad/scratch/PGS/data/derivatives/pgs"
pgs.ls <- data.frame(f = list.files(pgs.dir)) %>%
  mutate(PGS = sub(".rds", "", f)) %>%
  mutate(PGS = sub(".txt", "", PGS)) %>%
  filter(PGS != "PGI-ADHD")
pcs <- read_table("/Dedicated/jmichaelson-wdata/msmuhammad/scratch/PGS/data/derivatives/gcta_pc20_raw.eigenvec",
                  col_names = c("IID", "FID", paste0("PC", 1:20)))
registerDoMC(cores = 4)
pgs.all <- foreach(i=1:nrow(pgs.ls), .combine = rbind) %dopar% {
  raw <- read_rds(paste0(pgs.dir, "/", pgs.ls$f[i])) %>%
    mutate(pgs_name = sub(".txt", "", pgs_name)) %>%
    mutate(PGS= as.numeric(PGS))
  corrected <- inner_join(raw, pcs) %>%
    # mutate(corrected_pgs = PGS) %>%
    # mutate(corrected_pgs = scale(PGS, scale = T, center = T)[,1]) %>%
    mutate(corrected_pgs = residuals(glm(as.formula(paste0("PGS ~ ", paste(paste0("PC", c(1:20)), collapse = " + "))),
                                         data = inner_join(raw, pcs)))) %>%
    mutate(corrected_pgs = scale(corrected_pgs)[,1]) %>%
    select(IID, PGS, corrected_pgs, pgs_name)
  return(corrected)
}
write_rds(pgs.all, "/Dedicated/jmichaelson-wdata/msmuhammad/scratch/PGS/data/derivatives/all-pgs.rds")
```



### Notes

all data is aligned to hg19. Then, your genotypes were combined with EUR samples from the 1000 Genome project. Polygenic scores were then calculated using LDPRED2. The GWAS summary stats used for computing the PGS were from PGC and PGI mainly.

In the density plots below, I'm showing the distribution of the calculated PGS for all samples (i.e., YOU and the EUR sampled from the 1KG). I highlighted your computed PGS in a red dashed-line to show your PGS in reference to the population. 

A “Z-score” is a way to normalize data to have an average of 0, and a standard deviation of 1. Meaning a Z-score of -1, would be 1 standard deviation below average. This can easily be converted to a percentile (-1 z-score = 16th percentile, which means that in a room of 100 random people you would score higher than 16 of them).


```{r, fig.height=  9, fig.width=10}
pgs.all %>%
  ggplot(aes(x = corrected_pgs)) +
  geom_density() +
  facet_wrap(~pgs_name, scales = "free") +
  geom_blank(data = subset(pgs.all, IID == "you")) +
  geom_vline(data = subset(pgs.all, IID == "you"), aes(xintercept = corrected_pgs), 
             linetype = "dashed", color = "red")

```
The next barplot is showing the same data, but mainly focusing on your results. I'm using the Z-scores for this plot, and converted them to percentiles per PGS. I added horizpntal dashed-lines for percentiles of 25%, 50%, and 75%. 

50th percentile = average, > 50th percentile = above average, < 50th percentile = below average

```{r}
t <- pgs.all %>%
  select(-PGS) %>%
  group_by(pgs_name) %>%
  mutate(percentile = ecdf(corrected_pgs)(corrected_pgs)*100) %>%
  filter(IID == "you") %>%
  select(-IID, pgs_name, Z_score = corrected_pgs, percentile)
formatted_tbl(t)
```


```{r}
pgs.all %>%
  select(-PGS) %>%
  group_by(pgs_name) %>%
  mutate(percentile = ecdf(corrected_pgs)(corrected_pgs)*100) %>%
  filter(IID == "you") %>%
  ggplot(aes(x=pgs_name, y= percentile)) +
  geom_bar(stat = "identity", width = 0.5, fill = redblack.col[1]) +
  geom_hline(yintercept = c(25,50,75), linetype =2)
```

Same results are shown below, but with showing the Z-scores 

```{r}
pgs.all %>%
  select(-PGS) %>%
  # group_by(pgs_name) %>% mutate(corrected_pgs = scale(corrected_pgs, scale = T, center = T)[,1]) %>%
  filter(IID == "you") %>%
  ggplot(aes(x=pgs_name, y= corrected_pgs)) +
  geom_bar(stat = "identity", width = 0.5, fill = redblack.col[1]) +
  geom_hline(yintercept = c(-2,-1,1,2), linetype =2)
```




